core/lib/tee_verifier/src/lib.rs

```diff
@@ -0,0 +1,732 @@
+        system_env: SystemEnv,
+        used_contracts: Vec<(H256, Vec<u8>)>,
+    ) -> Self {
+        TeeVerifierInput::V1(V1TeeVerifierInput {
+            prepare_basic_circuits_job,
+            new_root_hash,
+            l2_blocks_execution_data: l2_blocks_execution_data
+                .into_iter()
+                .map(|v| v.into())
+                .collect(),
+            l1_batch_env: l1_batch_env.into(),
+            system_env: system_env.into(),
+            used_contracts,
+        })
+    }
+
+    /// Verify that the L1Batch produces the expected root hash
+    /// by executing the VM and verifying the merkle paths of all
+    /// touch storage slots.
+    ///
+    /// # Errors
+    ///
+    /// Returns a verbose error of the failure, because any error is
+    /// not actionable.
+    pub fn verify(self) -> anyhow::Result<()> {
+        let TeeVerifierInput::V1(V1TeeVerifierInput {
+            prepare_basic_circuits_job,
+            new_root_hash,
+            l2_blocks_execution_data,
+            l1_batch_env,
+            system_env,
+            used_contracts,
+        }) = self
+        else {
+            error!("TeeVerifierInput variant not supported");
+            bail!("TeeVerifierInput variant not supported");
+        };
+
+        let old_root_hash = l1_batch_env.previous_batch_hash.unwrap();
+        let l1_batch_number = l1_batch_env.number;
+        let l2_chain_id = system_env.chain_id;
+        let enumeration_index = prepare_basic_circuits_job.next_enumeration_index();
+
+        let mut raw_storage = InMemoryStorage::with_custom_system_contracts_and_chain_id(
+            l2_chain_id,
+            hash_bytecode,
+            Vec::with_capacity(0),
+        );
+
+        for (hash, bytes) in used_contracts.into_iter() {
+            trace!("raw_storage.store_factory_dep({hash}, bytes)");
+            raw_storage.store_factory_dep(hash, bytes)
+        }
+
+        let block_output_with_proofs =
+            Self::get_bowp_and_set_initial_values(prepare_basic_circuits_job, &mut raw_storage);
+
+        let storage_view = Rc::new(RefCell::new(StorageView::new(&raw_storage)));
+
+        let vm = VmInstance::new(l1_batch_env.into(), system_env.into(), storage_view);
+
+        let l2_blocks_execution_data = l2_blocks_execution_data
+            .into_iter()
+            .map(|v| v.into())
+            .collect();
+
+        let vm_out = Self::execute_vm(l2_blocks_execution_data, vm)?;
+
+        let instructions: Vec<TreeInstruction> =
+            Self::generate_tree_instructions(enumeration_index, &block_output_with_proofs, vm_out)?;
+
+        // `verify_proofs` will panic!() if something does not add up.
+        block_output_with_proofs
+            .verify_proofs(&Blake2Hasher, old_root_hash, &instructions)
+            .context("🛑 🛑 Failed to verify_proofs {l1_batch_number} correctly - oh no!")?;
+
+        if block_output_with_proofs.root_hash() != Some(new_root_hash) {
```

Comment:
Can this ever be true? If the proofs don't check out, the above statement will already bail out, no?

Reply:
Does not hurt to check.. right?

Reply:
ok, removed